File: Midterm Exam/cook.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <stdexcept>
#include <limits>

enum class Shift {
    DAY = 1,
    NIGHT = 2
};

class Date {
public:
    Date(int year, int month, int day) {
        if (year < 0) {
            throw std::invalid_argument("Year must be positive");
        }
        if (month < 1 || month > 12) {
            throw std::invalid_argument("Month must be between 1 and 12");
        }
        int days_in_month[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
        bool is_leap_year = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
        if (day < 1 || day > days_in_month[month - 1] || (month == 2 && day == 29 && !is_leap_year)) {
            throw std::invalid_argument("Day must be between 1 and the valid day for the month");
        }
        this->year = year;
        this->month = month;
        this->day = day;
    }

    bool operator<(const Date& other) const {
        if (year != other.year) return year < other.year;
        if (month != other.month) return month < other.month;
        return day < other.day;
    }

    std::string to_string() const {
        std::ostringstream oss;
        oss << year << "-" << month << "-" << day;
        return oss.str();
    }

private:
    int year;
    int month;
    int day;
};

class Cook {
public:
    Cook(const std::string& name, const Date& hire_date) : name(name), hire_date(hire_date) {}

    virtual ~Cook() {}

    Date getHireDate() const {
        return hire_date;
    }

    std::string getName() const {
        return name;
    }
    
protected:
    std::string name;
    Date hire_date;
};

class FastFoodCook : public Cook {
public:
    FastFoodCook(const std::string& name, const Date& hire_date, Shift shift)
        : Cook(name, hire_date), shift(shift) {}

    Shift getShift() const {
        return shift;
    }

private:
    Shift shift;
};

int main() {
    int num_cooks;
    std::cout << "Enter a number of fast food cooks: ";
    while (!(std::cin >> num_cooks)) {
        std::cout << "Invalid input. Try again: ";
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }

    std::vector<FastFoodCook> cooks;
    std::cin.ignore();
    for (int i = 0; i < num_cooks; ++i) {
        std::string name, hire_date_str;
        int shift_val, year, month, day;
        char dummy;

        std::cout << "Enter name of cook: ";
        std::getline(std::cin, name);

        std::cout << "Enter hire date of cook (YYYY-MM-DD): ";
        std::getline(std::cin, hire_date_str);
        std::istringstream iss(hire_date_str);
        if (!(iss >> year >> dummy >> month >> dummy >> day)) {
            std::cout << "Invalid date format. Try again." << std::endl;
            --i;
            continue;
        }

        std::cout << "Enter shift of cook (1 for day, 2 for night): ";
        if (!(std::cin >> shift_val) || (shift_val != 1 && shift_val != 2)) {
            std::cout << "Invalid shift. Try again." << std::endl;
            --i;
            continue;
        }

        cooks.push_back(FastFoodCook(name, Date(year, month, day), static_cast<Shift>(shift_val)));
        std::cin.ignore();
    }

    Date max_date(2019, 1, 1);
    std::cout << "Valid cooks:" << std::endl;
    for (const auto& cook : cooks) {
        if (cook.getShift() != Shift::DAY) {
            continue;
        }
        if (cook.getHireDate() < max_date) {
            std::cout << cook.getName() << std::endl;
        }
    }

}

File: Midterm Exam/cook.py
from enum import Enum

class Shift(Enum):
    DAY = 1
    NIGHT = 2

class Date:
    def __init__(self, year: int, month: int, day: int):
        if year < 0:
            raise ValueError("Year must be positive")
        if month < 1 or month > 12:
            raise ValueError("Month must be between 1 and 12")
        days_in_month = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31][month - 1]
        is_leap_year = year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
        if day < 1 or day > days_in_month and not (month == 2 and day == 29 and is_leap_year):
            raise ValueError("Day must be between 1 and " + str(days_in_month + (1 if month == 2 and is_leap_year else 0)))

        self.year = year
        self.month = month
        self.day = day
    
    def __str__(self):
        return str(self.year) + "-" + str(self.month) + "-" + str(self.day)
    
    def __lt__(self, other):
        if not isinstance(other, Date):
            raise TypeError("Cannot compare Date with " + str(type(other)))     
        if self.year < other.year:
            return True
        if self.year > other.year:
            return False
        if self.month < other.month:
            return True
        if self.month > other.month:
            return False
        if self.day < other.day:
            return True
        return False


class Cook:
    def __init__(self, name: str, hire_date: str):
        self.name = name
        self.hire_date: str = hire_date

class FastFoodCook(Cook):
    def __init__(self, name: str, hire_date: str, shift: Shift):
        super().__init__(name, hire_date)
        self.shift = shift
        
def main():
    valid = False
    while not valid:
        prompt = input("Enter a number of fast food cooks: ")
        try:
            num_cooks = int(prompt)
            valid = True
        except ValueError:
            print("Invalid input.")
    cooks: list[FastFoodCook] = []

    i = 0
    while i < num_cooks:
        name = input("Enter name of cook: ")
        hire_date = input("Enter hire date of cook: ") # Format: YYYY-MM-DD
        shift = input("Enter shift of cook (1 for day, 2 for night): ")
        try:
            shift = Shift(int(shift))
            hire_date = hire_date.split("-")
            hire_date = Date(int(hire_date[0]), int(hire_date[1]), int(hire_date[2])) 
        except ValueError:
            print("Invalid input.")
            continue
        cooks.append(FastFoodCook(name, hire_date, shift))
        i += 1

    max_date = Date(2019, 1, 1)

    valid_cooks = []

    for cook in cooks:
        if cook.shift != Shift.DAY:
            continue
        if cook.hire_date < max_date:
            valid_cooks.append(cook)
    
    print("Valid cooks:")
    for cook in valid_cooks:
        print(cook.name)

if __name__ == "__main__":
    main()